### PR TITLE
Application service room directories (and other third party network things)

### LINKED ISSUE #869: Spec per AS publicRooms list
(Depends on https://github.com/matrix-org/matrix-doc/issues/868)

AS's can publish rooms into their own public room list, so that they can advertise public rooms for their network without filling up the main public room list.

There are two sides to this:
1. the new client APIs for querying in /publicRooms
2. the new AS APIs that allow AS's to publish and unpublish rooms.

See https://github.com/matrix-org/synapse/pull/1676 for details

### PR DESCRIPTION
**Pull request by turt2live**
Rendered: see 'docs' status check.

This may be best to review commit by commit.

----

Move the third party network API to it's own section

Update third party network schemas
Some information was missed when this was reviewed. This commit adds some additional documentation for how these objects interact with each other.

Document third party network/protocol directories (for appservices)
Fixes #869

### CODE REVIEW

File: specification/application_service_api.rst [hunk 1]
```diff
@@ -180,6 +180,26 @@ this request (e.g. to join a room alias).
 {{query_room_as_http_api}}
 
 
+Third party networks
+++++++++++++++++++++
+
+Application services may declare which protocols they support via their registration
+file. These networks are generally for third party services such as IRC that the
```
Comment (ara4n) [inline specification/application_service_api.rst]: registration files are an implementation specific quirk, i think?  This should be "which protocols they support via their registration configuration at the homeserver"?

File: specification/application_service_api.rst [hunk 2]
```diff
@@ -180,6 +180,26 @@ this request (e.g. to join a room alias).
...
+application service is managing. Application services may populate a Matrix room
+directory for their registered protocols, as defined in the Client-Server API Extensions.
+
+Each protocol may have several "locations". A location within a protocol is a place
```
Comment (ara4n) [inline specification/application_service_api.rst]: for consistency with Leo's nomenclature elsewhere we should probably say "aka 3rd party locations or 3PLs")

File: api/application-service/protocols.yaml [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+            application/json: {
+              "errcode": "COM.EXAMPLE.MYAPPSERVICE_NOT_FOUND"
+            }
+          schema:
+            $ref: ../client-server/definitions/errors/error.yaml
+  "/_matrix/app/unstable/thirdparty/location":
+    get:
+      summary: Reverse-lookup third party locations given a Matrix room alias.
+      description: |-
+        Retreive an array of third party network locations from a Matrix room
```
Comment (ara4n) [inline api/application-service/protocols.yaml]: Retrieve

File: specification/application_service_api.rst (hunk 1 above)
Comment (turt2live) [reply specification/application_service_api.rst]: registration files are currently in the spec (introduced when the self-registration API was removed), although +1 to making it more generic for future support.

**Comment by ara4n** [comment]
(i guess this all looks okay, but i'm a bit worried about speccing stuff which might not have been used in anger)

**Comment by turt2live** [comment]
I believe the IRC bridge, Gitter, and Slack all use them (I think, at least going off of the riot room directory). Discord also makes use of them (I'm pretty sure).

**Comment by ara4n** [comment]
Hm. am I totally misreading this then? the API looks like it lets the HS query the AS for the list of rooms that the AS could expose - as opposed to just tagging which rooms happen to come from a given AS in the directory.  I'm aware we do the latter in publicRooms, but is anything doing the former?

File: specification/application_service_api.rst (hunk 1 above)
Comment (ara4n) [reply specification/application_service_api.rst]: oh, right, of course. prolly best leave it talking about 'registration file' or 'registration config file'.

**Comment by turt2live** [comment]
Ah, right. This isn't HS->AS communication (it's AS->HS with the intent of the HS taking care of it), but it is used at least by IRC and Gitter: https://github.com/search?q=org%3Amatrix-org+org%3Avector-im+setRoomDirectoryVisibilityAppService&type=Code

It also has sytests: https://github.com/matrix-org/sytest/blob/cab0f6b3a0ae8c8324f5c8ab609ac6286da178ca/tests/60app-services/06publicroomlist.pl#L53

**Comment by turt2live** [comment]
Although a search across all of github doesn't yield more than that, which may be a result of it not being in the spec?

File: api/application-service/protocols.yaml [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+          description: No users were found with the given parameters.
+          examples:
+            application/json: {
+              "errcode": "COM.EXAMPLE.MYAPPSERVICE_NOT_FOUND"
+            }
+          schema:
+            $ref: ../client-server/definitions/errors/error.yaml
+  "/_matrix/app/unstable/thirdparty/location/{protocol}":
+    get:
+      summary: Retreive Matrix-side portal rooms leading to a third party location.
```
Comment (ara4n) [inline api/application-service/protocols.yaml]: Retrieve

File: api/application-service/protocols.yaml [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+            application/json: {
+              "errcode": "COM.EXAMPLE.MYAPPSERVICE_NOT_FOUND"
+            }
+          schema:
+            $ref: ../client-server/definitions/errors/error.yaml
+  "/_matrix/app/unstable/thirdparty/user":
+    get:
+      summary: Reverse-lookup third party users given a Matrix User ID.
+      description: |-
+        Retreive an array of third party users from a Matrix User ID.
```
Comment (ara4n) [inline api/application-service/protocols.yaml]: Retrieve

**Comment by ara4n** [comment]
right, https://github.com/matrix-org/matrix-doc/commit/5403296 mitigates my confusion nicely; thanks. lgtm!

**Pull request merged**